src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php

```diff
@@ -219,54 +219,69 @@ private function autowireMethod(\ReflectionMethod $reflectionMethod, array $argu
                 continue;
             }
 
-            try {
-                if (!$typeHint = $parameter->getClass()) {
-                    // no default value? Then fail
-                    if (!$parameter->isOptional()) {
-                        if ($mustAutowire) {
-                            throw new RuntimeException(sprintf('Cannot autowire service "%s": argument $%s of method %s::%s() must have a type-hint or be given a value explicitly.', $this->currentId, $parameter->name, $reflectionMethod->class, $reflectionMethod->name));
-                        }
-
-                        return array();
-                    }
+            if (method_exists($parameter, 'getType')) {
+                if ($typeName = $parameter->getType()) {
+                    $typeName = $typeName->isBuiltin() ? null : ($typeName instanceof \ReflectionNamedType ? $typeName->getName() : $typeName->__toString());
+                }
+            } elseif (preg_match('/^(?:[^ ]++ ){4}([a-zA-Z_\x7F-\xFF][^ ]++)/', $parameter, $typeName)) {
```

Comment:
I would not use ``$typeName`` to store the regex matches. Repurposing the same variable name for something else makes the code much harder to read.

Same is true for the output of ``$parameter->getType()`` btw